Lib/unittest/test/test_case.py

```diff
@@ -1341,6 +1342,18 @@ class MyWarn(Warning):
             pass
         self.assertRaises(TypeError, self.assertWarnsRegex, MyWarn, lambda: True)
 
+    def testAssertWarnsModifySysModules(self):
+        # bpo-29620: handle modified sys.modules during iteration
+        class Foo(types.ModuleType):
+            @property
+            def __warningregistry__(self):
+                sys.modules['@bar@'] = 'bar'
+
+        sys.modules['@foo@'] = Foo('foo')
+        self.assertWarns(UserWarning, warnings.warn, 'expected')
```

Comment:
Does it cause a problem with other tests if this test fails and '@foo@/@bar@' are left in sys modules?  I wonder if there should be a `try...finally` here to ensure they get removed.